build.js

```diff
@@ -1824,10 +1861,26 @@ var dirtyDetectingBuilder = function(rootPackageName, resultsCache, prevResultsC
     var compileModulesCommands =
       singleFileCompile + compileModuleOutputs.map( function(c){return c.compileOutputString;}).join(' ');
 
+    var compileForeignDependencies = null;
```

Comment:
Did you test that this works when a *dependency* has a foreign dependency?
So you're compiling `A`, which depends on its `node_modules/B` and `B` has a `foreignDependency`.